src/Libraries/Microsoft.Extensions.AI.Abstractions/AdditionalUsageValues.cs

```diff
@@ -0,0 +1,349 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Runtime.InteropServices;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.AI;
+
+/// <summary>
+/// A key-value store for usage values. Values may be of the following types:
+/// <see cref="int"/>, <see cref="long"/>, <see cref="float"/>, <see cref="double"/>, <see cref="decimal"/>.
+/// </summary>
+[JsonConverter(typeof(AdditionalUsageValuesJsonConverter))]
+public class AdditionalUsageValues : IEnumerable<KeyValuePair<string, object>>
+{
```

Comment:
Are there other places we think we'll use this collection? I'm wondering if we'd be better off just using an `IDictionary<string, long>` for an `AdditionalCounts` property on UsageDetails, rather than trying to make it applicable to lots of different TValue types. Bonus, that would end up being cheaper for enumeration, since it wouldn't involve boxing.

Reply:
I'm not aware of other places this would be used, and yes I'd be happy to eliminate all this code.

The reason I didn't just go for `IDictionary<string, long>` is thinking there may be other summable value types people would want - floats and decimals in particular, e.g., to represent things like CPU/GPU time consumed and costs. Perhaps that's over-generalized and we can rely on people modelling things as whole-numbered `long` values anyway (e.g., GPU-microseconds and something like thousandths-of-cents or whatever smallest currency unit is needed).

Do you have a view on what's needed?

Reply:
Including `TimeSpan` on the discriminated union here would be fine, however it's not a huge benefit vs storing a nanosecond count as a `long` (especially since Ollama itself is documented to produce nanosecond counts). From my point of view this doesn't strongly push us one way or the other.

However I do agree we should move the timing info onto the `UsageDetails` so it can be aggregated. I'll update my PR to do that, once we've decided whether we're going with this one or #5709.